server/config.go

```diff
@@ -285,3 +295,159 @@ func (c *BalanceConfig) String() string {
+	AdvertisePeerURL    string `toml:"advertise-peer-url" json:"advertise-peer-url"`
+	AdvertiseClientURL  string `toml:"advertise-client-url" json:"advertise-client-url"`
+	InitialCluster      string `toml:"initial-cluster" json:"initial-cluster"`
+	InitialClusterState string `toml:"initial-cluster-state" json:"initial-cluster-state"`
+}
+
+const (
+	defaultEtcdName               = "default"
+	defaultEtcdDataDir            = "default.pd"
+	defaultEtcdWalDir             = ""
+	defaultEtcdListenPeerURL      = "http://localhost:2380"
+	defaultEtcdListenClientURL    = "http://localhost:2379"
+	defaultEtcdAdvertisePeerURL   = "http://localhost:2380"
+	defaultEtcdAdvertiseClientURL = "http://localhost:2379"
+	defaultEtcdInitialCluster     = "default=http://localhost:2380"
+)
+
+func newEtcdConfig() *EtcdConfig {
+	return &EtcdConfig{
+		Name:                defaultEtcdName,
+		DataDir:             defaultEtcdDataDir,
+		WalDir:              defaultEtcdWalDir,
+		ListenPeerURL:       defaultEtcdListenPeerURL,
+		ListenClientURL:     defaultEtcdListenClientURL,
+		AdvertisePeerURL:    defaultEtcdAdvertisePeerURL,
+		AdvertiseClientURL:  defaultEtcdAdvertiseClientURL,
+		InitialCluster:      defaultEtcdInitialCluster,
+		InitialClusterState: embed.ClusterStateFlagNew,
+	}
+}
+
+func (c *EtcdConfig) clone() *EtcdConfig {
+	if c == nil {
+		return nil
+	}
+
+	ec := &EtcdConfig{}
+	*ec = *c
+	return ec
+}
+
+// String implements fmt.Stringer interface.
+func (c *EtcdConfig) String() string {
+	if c == nil {
+		return "<nil>"
+	}
+
+	return fmt.Sprintf("EtcdConfig(%+v)", *c)
+}
+
+// generates a configuration for embedded etcd.
+func (c *Config) genEmbedEtcdConfig() (*embed.Config, error) {
+	cfg := embed.NewConfig()
+	cfg.Name = c.EtcdCfg.Name
+	cfg.Dir = c.EtcdCfg.DataDir
+	cfg.WalDir = c.EtcdCfg.WalDir
+	cfg.InitialCluster = c.EtcdCfg.InitialCluster
+	// Use unique cluster id as the etcd cluster token too.
+	cfg.InitialClusterToken = fmt.Sprintf("pd-%d", c.ClusterID)
+	cfg.ClusterState = embed.ClusterStateFlagNew
+	cfg.EnablePprof = true
+
+	u, err := url.Parse(c.EtcdCfg.ListenPeerURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.LPUrls = []url.URL{*u}
+
+	u, err = url.Parse(c.EtcdCfg.ListenClientURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.LCUrls = []url.URL{*u}
+
+	u, err = url.Parse(c.EtcdCfg.AdvertisePeerURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.APUrls = []url.URL{*u}
+
+	u, err = url.Parse(c.EtcdCfg.AdvertiseClientURL)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg.ACUrls = []url.URL{*u}
+
+	return cfg, nil
+}
+
+// Generate a unique port for etcd usage. This is only used for test.
+// Because initializing etcd must assign certain address.
+func freePort() int {
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		panic(err)
+	}
+
+	defer l.Close()
+
+	return l.Addr().(*net.TCPAddr).Port
+}
```

Comment:
if we use this port quickly, it won't . because system will wait some time to reuse this port again.

I google it and find all use this way. so I think it's ok in test.